syntax = "proto3";

package blockchain;

message ConnectMessage {
    int32 update = 1;
    int32 verify = 2;
    int32 client_id = 3;
    Chain chain = 4;

    int32 mine = 5;
    int32 miner_id = 6;
}

message MineMessage {
    int32 difficulty = 1;
    string hash = 2;
}

message NonceMessage {
    int32 nonce = 1;
    int32 miner_id = 2;
}

message UpdateMessage {
    enum UpdateType {
        ENTRY = 0;
        BLOCK = 1;
    }

    UpdateType type = 1;
    Entry entry = 2;
    Block block = 3;
}

message Entry {
    string sender = 1;
    string receiver = 2;
    int32 amount = 3;
}

message Block {
    string prev_hash = 1;
    repeated Entry entries = 3;
}

message Chain {
    int32 block_size = 1;
    repeated Block blocks = 2;
}